### PR TITLE
Fixing several problems with hypothes.extra.django

### PR DESCRIPTION
**Pull request by etianen**
I really hate it when people submit a monster pull request to my projects. So... I decided to submit a monster pull request to yours.

This improves `hypothesis.extra.django` in the following ways:

- Allows developers to override the default field mappings using `add_default_field_mapping` (previously impossible).
- Performs a `full_clean()` on all models before saving, ensuring they're valid and allowing the model to perform it's full lifecycle.
- Provides public strategies for generating data for many django model field types. For example, you can get data for a `SlugField` using `slug_field_values(MyModel._meta.get_field("some_field"))`
- Provides a public strategy for generating data for any registered field `field_values(MyModel._meta.get_field("some_field"))`
- Provides a `defines_field_strategy` decorator for easily creating new field strategies.
- Minimizes the size of generated data by biasing towards model default fields, configurable via the `__default_bias` argument to `models()`. Non-trivial models previously overflowed the data buffer frequently, triggering health checks.
- Added built-in support for `UrlField, DateField, TimeField` and `SlugField` mapping.
- Multi-db support for `models()` strategy via `__db` parameter.
- Ability to use `add_default_field_mapping` as a decorator to a strategy factory.

Fixes bugs:
- Uncaught `DataErrors` no longer blowing up tests.
- Fields no longer silently truncated by unicode null.
- `EmailField` strategy no longer generates values that don't pass `EmailValidator`.

Potential "breaking" changes:
- The `default_value` strategy used to omit values from the model data dict. Now it includes the field's default value. This is now in line with what the public docs actually say, but it's still different from current behaviour. I'd class this as a bugfix.

To do:

- Improve performance of UrlField and EmailField implementations.
- Update documentation.
- Fix formatting errors.

I'll happily tackle these todos if I get some sort of positive feedback on these changes. :)

### CODE REVIEW

File: src/hypothesis/extra/django/models.py [hunk 1]
```diff
@@ -15,161 +15,303 @@
...
 from hypothesis.errors import InvalidArgument
 from hypothesis.extra.datetime import datetimes
+from hypothesis.extra.fakefactory import fake_factory
 from hypothesis.utils.conventions import UniqueIdentifier
-from hypothesis.searchstrategy.strategies import SearchStrategy
 
 
-class ModelNotSupported(Exception):
-    pass
-
```
Comment (etianen) [inline src/hypothesis/extra/django/models.py]: This exception wasn't raised anywhere, so I killed it.

File: src/hypothesis/extra/django/models.py [hunk 2]
```diff
@@ -15,161 +15,303 @@
...
-def referenced_models(model, seen=None):
-    if seen is None:
-        seen = set()
-    for f in model._meta.concrete_fields:
-        if isinstance(f, dm.ForeignKey):
-            t = f.rel.to
-            if t not in seen:
-                seen.add(t)
-                referenced_models(t, seen)
-    return seen
```
Comment (etianen) [inline src/hypothesis/extra/django/models.py]: This wan't used anywhere or documented, so I killed it.

File: src/hypothesis/extra/django/models.py [hunk 3]
```diff
@@ -17,159 +17,332 @@
...
             data.mark_invalid()
+        except (ValidationError, IntegrityError):
+            # This might mean that a unique key way violoated.
+            # As a fallback, try to get the identical model. This is needed
+            # for example() calls, which expect to be able to call the
+            # strategey idempotently.
+            try:
+                return self.model._default_manager.get(**model_data)
+            except self.model.DoesNotExist:
+                # Something other than a unique violoation
```
Comment (alexwlchan) [inline src/hypothesis/extra/django/models.py]: Minor: "way violoated"? Did you mean "was violated"?

In the same comment: "strategey" ~> "strategy" and "violoation" ~> "violation".

**Comment by etianen** [comment]
Apologies for spamming travis with commits. The `make format` command, plus friends just error on my dev laptop, so I've got no choice but to let Travis do the lifting for me. :(

```
make format
scripts/retry.sh scripts/install.sh 3.4
+ env
+ grep UTF
LC_ALL=en_US.UTF-8
LANG=en_GB.UTF-8
+ BASE=/Users/dave/.cache/hypothesis-build-runtimes
+ mkdir -p /Users/dave/.cache/hypothesis-build-runtimes
+ LOCKFILE=/Users/dave/.cache/hypothesis-build-runtimes/.install-lockfile
+ true
+ mkdir /Users/dave/.cache/hypothesis-build-runtimes/.install-lockfile
+ echo 'Successfully acquired installer.'
Successfully acquired installer.
+ break
+ trap 'rm -rf /Users/dave/.cache/hypothesis-build-runtimes/.install-lockfile' EXIT
+ PYENV=/Users/dave/.cache/hypothesis-build-runtimes/pyenv
+ '[' '!' -d /Users/dave/.cache/hypothesis-build-runtimes/pyenv/.git ']'
+ back=/Users/dave/Workspace/hypothesis
+ cd /Users/dave/.cache/hypothesis-build-runtimes/pyenv
+ git fetch
+ git reset --hard origin/master
HEAD is now at a300254 Added new pythons to changelog
+ cd /Users/dave/Workspace/hypothesis
+ SNAKEPIT=/Users/dave/.cache/hypothesis-build-runtimes/snakepit
+ for var in '"$@"'
+ case "${var}" in
+ install 3.4.3 python3.4
+ VERSION=3.4.3
+ ALIAS=python3.4
+ mkdir -p /Users/dave/.cache/hypothesis-build-runtimes/versions
+ SOURCE=/Users/dave/.cache/hypothesis-build-runtimes/versions/python3.4
+ '[' '!' -e /Users/dave/.cache/hypothesis-build-runtimes/versions/python3.4 ']'
+ rm -f /Users/dave/.cache/hypothesis-build-runtimes/snakepit/python3.4
+ mkdir -p /Users/dave/.cache/hypothesis-build-runtimes/snakepit
+ /Users/dave/.cache/hypothesis-build-runtimes/versions/python3.4/bin/python -m pip.__main__ install --upgrade pip wheel virtualenv
scripts/install.sh: line 81: /Users/dave/.cache/hypothesis-build-runtimes/versions/python3.4/bin/python: No such file or directory
+ rm -rf /Users/dave/.cache/hypothesis-build-runtimes/.install-lockfile
Command failed. Retrying...
+ env
+ grep UTF
LC_ALL=en_US.UTF-8
LANG=en_GB.UTF-8
+ BASE=/Users/dave/.cache/hypothesis-build-runtimes
+ mkdir -p /Users/dave/.cache/hypothesis-build-runtimes
+ LOCKFILE=/Users/dave/.cache/hypothesis-build-runtimes/.install-lockfile
+ true
+ mkdir /Users/dave/.cache/hypothesis-build-runtimes/.install-lockfile
+ echo 'Successfully acquired installer.'
Successfully acquired installer.
+ break
+ trap 'rm -rf /Users/dave/.cache/hypothesis-build-runtimes/.install-lockfile' EXIT
+ PYENV=/Users/dave/.cache/hypothesis-build-runtimes/pyenv
+ '[' '!' -d /Users/dave/.cache/hypothesis-build-runtimes/pyenv/.git ']'
+ back=/Users/dave/Workspace/hypothesis
+ cd /Users/dave/.cache/hypothesis-build-runtimes/pyenv
+ git fetch
+ git reset --hard origin/master
HEAD is now at a300254 Added new pythons to changelog
+ cd /Users/dave/Workspace/hypothesis
+ SNAKEPIT=/Users/dave/.cache/hypothesis-build-runtimes/snakepit
+ for var in '"$@"'
+ case "${var}" in
+ install 3.4.3 python3.4
+ VERSION=3.4.3
+ ALIAS=python3.4
+ mkdir -p /Users/dave/.cache/hypothesis-build-runtimes/versions
+ SOURCE=/Users/dave/.cache/hypothesis-build-runtimes/versions/python3.4
+ '[' '!' -e /Users/dave/.cache/hypothesis-build-runtimes/versions/python3.4 ']'
+ rm -f /Users/dave/.cache/hypothesis-build-runtimes/snakepit/python3.4
+ mkdir -p /Users/dave/.cache/hypothesis-build-runtimes/snakepit
+ /Users/dave/.cache/hypothesis-build-runtimes/versions/python3.4/bin/python -m pip.__main__ install --upgrade pip wheel virtualenv
scripts/install.sh: line 81: /Users/dave/.cache/hypothesis-build-runtimes/versions/python3.4/bin/python: No such file or directory
+ rm -rf /Users/dave/.cache/hypothesis-build-runtimes/.install-lockfile
Command failed. Retrying...
```

File: src/hypothesis/extra/django/models.py (hunk 3 above)
Comment (etianen) [inline src/hypothesis/extra/django/models.py]: Gooood spot. Fixed in another commit.

**Comment by DRMacIver** [comment]
Wow, thanks for this!

It might take a little while for me to review it, as I'm off on a trip shortly and have a bunch to do before then, but I am very keen on it.

RE make format not working: That's weird. Do you have a $HOME/.cache? I guess the script is assuming that it exists and not creating it if it doesn't. I'll fix if that's the problem.

File: src/hypothesis/extra/django/models.py [hunk 4]
```diff
@@ -17,159 +17,332 @@
...
+                ]
+                if field.blank:
+                    choices.extend(blank_choices)
+                strategy = st.sampled_from(choices)
+            else:
+                # Run the factory function.
+                strategy = func(field, **kwargs)
+            # Add in null values.
+            if field.null:
+                strategy = st.one_of(strategy, st.none())
```
Comment (DRMacIver) [inline src/hypothesis/extra/django/models.py]: Due to technical reasons of how shrinking works, this is better written `st.one_of(st.none(), strategy)` (shrinking will try to shrink towards the left of one_of)

**Comment by etianen** [comment]
I use hypothesis for a lot of projects, and the Django support was bugging me for being *almost* incredibly useful, but requiring a lot of per-project tweaks to make work for real models.

Although it's breaking, I'd also suggest making the new "minimal" mode default. Any non-trivial model routinely blows up the data buffer by requiring all fields to be filled. In my current project, it's either minimal mode or 32K buffers!

Meanwhile, I'm gonna aim for green across the board with the CI tests, and update the docs when I have some time. Enjoy the trip!

(I do have a `.cache`, yeah. I'll have a go at debugging what's up if I have time, but it's easy for now to just let Travis do my work for me!)

**Comment by DRMacIver** [comment]
> I use hypothesis for a lot of projects, and the Django support was bugging me for being almost incredibly useful, but requiring a lot of per-project tweaks to make work for real models.

Thanks, this is useful feedback. I don't use Django much myself, so it's hard for me to seethings like this.

> Although it's breaking, I'd also suggest making the new "minimal" mode default. 

I don't think I'd count this as breaking (it's certainly API compatible!), but I wonder if there is some sort of useful halfway house here where the behaviour is that of the minimal one but there's something you can pass as a value that means "use the full range".

Another possible option is just to include a strong bias towards using the default where one is set. e.g. if you have N fields with default set, each one could be set to use a non-default value with probability 1/N. That way the expected amount of data used by a field with lots of models with default values is low and you don't have to manually fine-tune this. How does that sound?

> Any non-trivial model routinely blows up the data buffer by requiring all fields to be filled. In my current project, it's either minimal mode or 32K buffers!

Hmm. That's definitely not ideal behaviour, especially given how slow the Django testing already is!

> (I do have a .cache, yeah. I'll have a go at debugging what's up if I have time, but it's easy for now to just let Travis do my work for me!)

Sure, no worries. You're definitely welcome to do that. I'm just puzzled because the Makefile definitely does work on OSX because I have travis builders on OSX, so if you do have a moment to debug that would be appreciated.

**Comment by etianen** [comment]
You can already toggle the behaviour with the `_minimal=False` argument to `models()`. It would just be a case of flipping it to `_minimal=True`, really.

Adding a strong bias to using default values is even better. I just haven't the faintest clue how to do this! I've searched the docs for "bias" and "distribution" to no avail. I could do something using `randoms()`, and use the default value when random < 0.1 or something. It depends if there's a built-in way of doing this more nicely.

**Comment by DRMacIver** [comment]
> You can already toggle the behaviour with the _minimal=False argument to models(). It would just be a case of flipping it to _minimal=True, really.

Well, what I had in mind was being able to toggle it on a per field basis.

> Adding a strong bias to using default values is even better. I just haven't the faintest clue how to do this! I've searched the docs for "bias" and "distribution" to no avail. I could do something using randoms(), and use the default value when random < 0.1 or something. It depends if there's a built-in way of doing this more nicely.

Yeah, the distribution stuff is all a bit internal and undocumented for a variety of reasons, mostly it tends to be subject to change.

The easiest thing to do to get a 1/N bias is going to be something like `integers(1, N).flatmap(lambda i: just(default) if i < N else field_strategy)`

**Comment by etianen** [comment]
So how about this:

The default behaviour is that non-mandatory fields have a 0.9 chance of being default. This is configured by a `_default_bias=0.9`  flag to `models()`.

You can override this on a per-field basis by simply supplying the built-in search strategy factory to the field overrides: `models(MyModel, my_field=field_values)`, which in this case would use `field_values(MyModel._meta.get_field("my_field"))` as the strategy.

**Comment by DRMacIver** [comment]
I think having the bias depend on the number of fields with defaults is probably a good idea, because it means that the size of the data stays constant on average as the model size grows.

Also, somewhat frustratingly, adding new arguments to the models function would require a major version bump, so I'd rather you don't do that.

**Comment by etianen** [comment]
Unfortunately, I've already done that with the multi-db support. What do you suggest here? Take out multi-db support (sad)? Another function that provides this extra functionality, and have models() wrap it (cludgy)?

**Comment by DRMacIver** [comment]
Hmm. How about starting additional argument names with \_\_ instead of just \_? It's a bit ugly but I think I can validly argue that that's not a breaking change because any fields whose name starts with __ would get named mangled anyway, so you couldn't have previously used those arguments.

**Comment by etianen** [comment]
Yes, the double underscore makes sense here. It's annoying adapting an API that previously had free range to use **kwargs!

**Comment by etianen** [comment]
Okay, I've sorta implemented the default bias feature as described.

Varying the default bias according to the number of fields proved tricky, so it's currently not implemented. Mostly this is in terms of making the feature understandable to the end user. When it doesn't vary according to fields, you can say:

- 0.0 means 0% of non-mandatory fields will have default values.
- 0.5 means 50% of non-mandatory fields will have default values.
- 1.0 means 100% of non-mandatory fields will have default values.

Which is nice and simple. If it varies according to fields, then the description becomes more like:

- 0.5 means 50% of models will have an average of one non-default value in their non-mandatory fields.
- 1.0 means 100% of models will have an average of one non-default value in their non-mandatory fields?
- 2.0 means 100% of models will have an average of two non-default value in their non-mandatory fields?

It starts making it hard for people to reason instinctively about the examples they're getting. It's probably easier to just add a hint of "if your models are giving too_large health checks, try increasing the value of __default_bias".

Incidentally, this has introduced a potentially breaking change to some edge-case behaviour, documented in the description of this pull request.

Oh, and I haven't touched tests yet. It's good to make sure that the current changed don't break any of the existing tests in a meaningful way, but I'd actually like to rewrite the tests to simplify them down and target the bits of the code that actually need stress-testing.

**Comment by DRMacIver** [comment]
I don't think having a manually configurable bias is a useful thing at all TBH. I think the desired behaviour is to have models() do something sensible by default and then if people want a different behaviour for specific fields override those. I'm happy to be convinced otherwise, but I don't think there's a useful constant value you can put there that would make a good default.

**Comment by etianen** [comment]
I guess the reason to expose it as a constant value is:

1. If we're going to bias fields towards their default value, we need a measure of how biased that will be, which is currently 90% of examples will have the default value (0.9).
2. So we have a constant of 0.9 in the codebase.
3. Some users will want to tweak this to make, say, no fields generate the default value, or all fields generate the constant value. It lets you balance randomness of data versus slowness of test.
4. So we let them do so.

Also, some users are gonna want to restore the "old" behaviour, since it's fine for models with only a few low-complexity fields.

**Comment by DRMacIver** [comment]
No, sorry, I'm still unconvinced. The parameter you're suggesting literally has no correct default value, which means that everyone is going to have to set it or suffer, while at the same time its existence significantly constrains future options for doing the right thing automatically.

In fact the more I think about it the more I think this is the wrong place to solve this. I think with a little bit of careful design Hypothesis can detect this sort of situation automatically and help work around it, and the correct thing to do here is just going to be to make fields with a default be something like `one_of(default_value(field), field_value(field))` and have Hypothesis be able to do the biasing automatically for things like that.

**Comment by DRMacIver** [comment]
(Obviously I don't expect you to do the "Work things out automatically" part. I think for now I'm just going to ask you to pull the minimality stuff from this PR and I'll sort out that logic in paralle)

**Comment by etianen** [comment]
I accept your suggestion that I do less work. :)

So, do you want an explicit one_of(default_value) call to help your biasing
work? It's not actually needed for the Django stuff, since the field
strategies I have will already generate default values automatically.

Oh, and as an aside, is there any utility in me creating a custom
SearchStrategy class rather than just using the composite decorator? I was
just following how the previous implementation did it.

**Comment by lucaswiman** [comment]
> you can no longer count on default field mappings of just(something) only generating that value for fields with a default value (90% of the time you'll get the default value back). 

Is there a way to set the default bias to 0 for `just`? This seems like very counterintuitive behavior that changes the way `just` works. At the least if the behavior of `just` changes, it should be noted in the docs for both extras.django and `just`.

File: src/hypothesis/extra/django/models.py [hunk 5]
```diff
@@ -17,159 +17,402 @@
...
+
+    - Fields with choices.
+    - Null fields.
+    - Filtering by field validators.
+
+    """
+    def decorator(func):
+        @defines_strategy
+        @wraps(func)
+        def create_field_strategy(field, **kwargs):
```
Comment (lucaswiman) [inline src/hypothesis/extra/django/models.py]: This is really nice. 👍

File: src/hypothesis/extra/django/models.py [hunk 6]
```diff
@@ -17,159 +17,402 @@
...
+                ]
+                if field.blank:
+                    choices.extend(blank_values)
+                strategy = st.sampled_from(choices)
+            else:
+                # Run the factory function.
+                strategy = func(field, **kwargs)
+            # Add in null values.
+            if field.null:
+                strategy = st.one_of(st.none(), strategy)
```
Comment (lucaswiman) [inline src/hypothesis/extra/django/models.py]: In cases where the field is nullable but has a `choices` attribute, this will strongly bias towards null values. Would it make sense to add `None` to the choices list passed to `sampled_from`?

File: src/hypothesis/extra/django/models.py [hunk 7]
```diff
@@ -17,159 +17,402 @@
...
+
+
+@add_default_field_mapping(dm.CharField)
+@add_default_field_mapping(dm.TextField)
+@defines_field_strategy(blank_values=('',))
+def char_field_values(field, **kwargs):
+    """A strategy of valid values for the given CharField."""
+    if field.blank:
+        kwargs.setdefault('min_size', 0)
+    if field.max_length:
```
Comment (lucaswiman) [inline src/hypothesis/extra/django/models.py]: Nit: is it possible to have a field with `max_length == 0`? Maybe update this to `if field.max_length is not None:`

File: src/hypothesis/extra/django/models.py [hunk 8]
```diff
@@ -17,159 +17,402 @@
...
+                if not isinstance(
+                    self.model._meta.get_field(field_name),
+                    dm.AutoField,
+                )
+            }
+            if model_keys:
+                try:
+                    return self.model._default_manager.get(**model_keys)
+                except self.model.DoesNotExist:
+                    # It was something other than a unique violation
```
Comment (lucaswiman) [inline src/hypothesis/extra/django/models.py]: Would it make sense to handle `MultipleObjectsReturned` by choosing among the matching rows?

File: src/hypothesis/extra/django/models.py (hunk 6 above)
Comment (etianen) [inline src/hypothesis/extra/django/models.py]: This has been done in the latest set of commits. Good call!

File: src/hypothesis/extra/django/models.py (hunk 7 above)
Comment (etianen) [inline src/hypothesis/extra/django/models.py]: Yeah, you're right. I've handled this edge case.

File: src/hypothesis/extra/django/models.py (hunk 8 above)
Comment (etianen) [inline src/hypothesis/extra/django/models.py]: Good point. Updated the pull request.

**Comment by etianen** [comment]
Okey-dokey, I've pulled out the default_bias/mimimal models stuff, replaced with adding the default value to a one_of() strategy to bias the results to that strategy. In the new tests, which specifically test a very large model, this seems to be handling the buffer overflow issues we were having before.

Speaking of new tests, I've refactored the tests in a way I hope you will find pleasing. It covers all the new functionality, plus the old, I believe.

@lucaswiman - I've included your suggestions in the latest pull request.

However, I've had to disable the slow health checks for the tests, for now, as data generation seems pretty slow on my ageing laptop. From my investigation, this seems tied to the EmailField and UrlField mappings relying on fake_factory, which seems to produce data quite slowly. I was wondering about writing some dumb implementations of strategies for these field types to drop the fake_factory dependency. Kinda like #162. For now, these could just sit in the Django extras package, as they probably won't pass muster for a general email/url generation strategy, but they will allow the Django integration to run faster.


There's probably formatting errors everywhere, I'll fix that when the code is otherwise done.

**Comment by DRMacIver** [comment]
Yeah, I very much want to ditch the dependency on fake factory for the django integration. It's both very slow and not very good.

> So, do you want an explicit one_of(default_value) call to help your biasing
work? It's not actually needed for the Django stuff, since the field
strategies I have will already generate default values automatically.

Yeah, i do want this. Even without the biasing work it will probably help keep the size down.

> Oh, and as an aside, is there any utility in me creating a custom
SearchStrategy class rather than just using the composite decorator? 

Not much. There used to be, but the difference has largely gone away.

**Comment by etianen** [comment]
Okay, next week I'm happy to have a go at reverse-engineering the regex validators in Django to make a fairly comprehensive URL and email generator strategy. I guess these would ideally be a part of the core strategies for hypothesis?

**Comment by lucaswiman** [comment]
One solution to both problems would be to implement https://github.com/HypothesisWorks/hypothesis-python/issues/170 for regular grammars, since URLs, IP addresses, email addresses can be expressed as regular expressions without too much difficulty.

**Comment by DRMacIver** [comment]
Yeah, that's a current work in progress, but it's a bit stalled due to reasons. For now I'd be happy to accept a manually tuned one and then if and when that arrives they can be replaced.

**Comment by etianen** [comment]
So, I got a bit carried away... and implemented #170, #162 and #267

The short of it is that you can now to this:

``` py
from hypothesis import given, strategies as st, patterns as p

@given(st.patterns(p.urls))
def test_urls(value):
    assert value

@given(st.patterns(p.emails))
def test_emails(value):
    assert value
```

More specifically, there's a new strategy called `patterns(pattern, min_size=None, max_size=None)`, that is used to generate strings matching a pattern.

So what is a pattern? Why, it's this adorable DSL I made!

``` py
davids = p.just("David") + p.just(" ") + p.one_of("MacIver", "Hall")
# => "David Hall" or "David MacIver"

too_many_davids = davids.repeat(min_repeat=3).join(p.just(" ")
# => "David Hall David MacIver David Hall David MacIver David MacIver"
```

The following pattern primitives are supported, all of which work like the equivalently-named strategies.

* `text(alphabet=None, min_size=None, max_size=None)`
* `just(str)`
* `one_of(*patterns)`
* `maybe(pattern)`
* `sampled_from([str])`

The following operators are supported:

* `__or__`
* `__add__`
* `limit(min_size=None, max_size=None)`
* `repeat(min_repeat=None, max_repeat=None)`
* `__mul__`
* `join(pattern)`

The URL and email generator both pass Django validation, and are quite nasty. For example, here is a perfectly cromulent URL:

ftps://²핁2৻왩즏഻隓ī鹧Ã꼠*@锉엀Ą❘À???ÅǸè«ãƢ¬䇃öǔᙆ䏛劜츂į妆ĕȚ¿揶èĄ.drH??????.rĶ

``` py
>>> from urllib.parse import urlparse
>>> urlparse("ftps://²핁2৻왩즏഻隓ī鹧Ã꼠*@锉엀Ą❘À???ÅǸè«ãƢ¬䇃öǔᙆ䏛劜츂į妆ĕȚ¿揶èĄ.drH??????.rĶ")
ParseResult(scheme='ftps', netloc='²핁2৻왩즏\u0d3b隓ī鹧Ã꼠*@锉엀Ą❘À', path='', params='', query='??\ue92dÅǸè«ãƢ¬䇃öǔᙆ䏛劜츂į妆ĕȚ¿\uea13揶èĄ.drH??????.rĶ', fragment='')
```

Things that URLs can include:

* ftp, ftps, http, https schemes
* Optional URL auth
* domain names, ipv4 and ipv6 addresses
* Optional port
* Optional path
* Optional query string
* Optional fragment

One downside - the URL generation strategy is currently quite slow, due to the huge complexity of potential URLs. Some advice on further optimization would be appreciated here, or maybe I just have to reduce the complexity of the URL pattern (and thus the nastiness, but it's a compromise between performance and nastiness).

And, to forstall some obvious questions:

**Q: Why not make a rexex decompiler instead, or some other text-based grammar?**

It would be even more complicated that this, which was quite complicated!


**Q: Why not build this on top of SearchStrategy instead of a whole new zoo of pattern primitives and operators?**

1. Correctly applying min and max size bounds to a pattern means that the min and max size of each part of the pattern's grammar tree needs to be propagated outwards. Which turns out is quite an invasive change.
2. Correctly applying min and max size bounds also requires an API for asking a pattern to resize itself within it's allowed bounds based on outside pressure. 
3. The normal strategies operate on all sorts of values, not just strings, so a lot of type checking would be involved.
4. Implementing a cute DSL is pretty important for building up complex patterns, which means more invasive changes to SearchStrategy.
5. In order to make it remotely fast, a certain amount of branch flattening optimisations had to be done to the patterns, which involves more invasive changes.


Things to do:

* Formatting does not match project standard!
* There are lots of tests, but I haven't checked full coverage of patterns yet (I'm sleepy!)
* Code review by other people! I'm happy with what I've got here, but there are probably many improvements!

File: src/hypothesis/patterns.py [hunk 9]
```diff
@@ -0,0 +1,549 @@
...
+    def __init__(self, value):
+        super(JustPattern, self).__init__(len(value), len(value))
+        self._value = value
+        self._strategy = st.just(self._value)
+
+    def _get_strategy_impl(self, min_size, max_size):
+        return self._strategy
+
+
+def just(value):
```
Comment (lucaswiman) [inline src/hypothesis/patterns.py]: This is a name conflict with the `just` strategy. What do you think about calling this `literal`?

File: src/hypothesis/patterns.py [hunk 10]
```diff
@@ -0,0 +1,549 @@
...
+    return JustPattern(_expect_text(value))
+
+
+def sampled_from(values):
```
Comment (lucaswiman) [inline src/hypothesis/patterns.py]: Also a name conflict. I think you could simplify the API here by allowing `one_of` to accept either string literals or other strategies, something like:
```python
def one_of(*patterns):
    patterns = [just(p) if isinstance(p, basestring) else p for p in patterns]
    return reduce(or_, patterns)
```

**Comment by lucaswiman** [comment]
I suspect it will be fairly difficult to adapt this approach to get a uniform distribution over strings matching the regular expression with this approach, but this would definitely be very useful!

> Q: Why not make a rexex decompiler instead, or some other text-based grammar?
> 
> It would be even more complicated that this, which was quite complicated!

I've been working on a project to do example generation from a regular expression [here](https://github.com/lucaswiman/revex/blob/7323f2c94da1463d18fee1fb861fd9e52e206602/revex/machine.py#L532).  I agree that it would be fairly complex to do this, even if you used the undocumented `re.sre_parse.parse` method that would do the parsing for you.

**Comment by DRMacIver** [comment]
Sorry, but your pattern work overlaps and conflicts with a lot of quite specific stuff I have planned and am already working on 

I've not done a review because I'm replying between moments while on holiday but I wanted to make sure to head you off before you sunk a lot of work into it.

I really do appreciate the work you're doing, but I'd appreciate it a lot more if you kept it a bit more focused rather than trying to implement all the things in one pull request.

**Comment by etianen** [comment]
I totally understand.

I'd maybe suggest that implementing a decent URL and email generator
without accidentally creating a pattern strategy is about as complicated as
simply creating a pattern strategy! A decent implementation of urls() would
have been monstrous by any measure!. :P

So... I guess that leaves the options of:

1. Keep Django implementation dependent on fake factory, and thus very slow
for non-trivial models. This is my favoured approach now.

2. Make a half-arsed URL and email strategy, that is private to Django, and
accept the poor data.

I'll pull the pattern generation stuff into a standalone project, since it
works well and I need something like this now. (I figured out the
optimisation issues last night.)

Does seem a shame though. There's several tickets asking for this stuff.

Lucas - The names were chosen deliberately to match existing strategies,
since they work in an analogous way, but specialised for text. The hope
being to improve familiarity (although such discussion is now irrelevant!)

**Comment by DRMacIver** [comment]
>  Keep Django implementation dependent on fake factory, and thus very slow
for non-trivial models. This is my favoured approach now.

I think this is the best option for now.

> Does seem a shame though. There's several tickets asking for this stuff.

Yes, and they'll get it, just not yet :-)

File: src/hypothesis/extra/django/models.py [hunk 11]
```diff
@@ -17,159 +17,359 @@
...
+    field_strategies = {
+        field_name: _resolve_strategy_factory(
+            strategy,
+            model._meta.get_field(field_name),
+        )
+        for field_name, strategy
+        in field_strategies.items()
+    }
+    # Introspect model for extra fields.
+    for field in model._meta.concrete_fields:
```
Comment (gcbirzan) [inline src/hypothesis/extra/django/models.py]: This should be changed to a .get_fields() call with a filter for f.is_concrete, concrete_fields is an [internal API](https://github.com/django/django/blob/master/django/db/models/options.py#L501).

Comment (etianen) [inline src/hypothesis/extra/django/models.py]: It's also not part of Django 1.7, which hypothesis supports.

**Comment by etianen** [comment]
I'm afraid I've lost interest with this. Feel free to use this or not as you feel best. I'm unsubscribing from notifications.
